js/api.ts

```diff
@@ -0,0 +1,49 @@
+import { fetch, reload } from "./browser";
+import { useEffect, useState } from "react";
+import { z } from "zod";
+
+export type ApiResult<Result> =
+  | { status: "loading" }
+  | { status: "ok"; result: Result }
+  | { status: "error" };
+
+export const useApiResult = <RawData, Data>({
+  RawData,
+  url,
+  parser,
+}: {
+  RawData: z.ZodType<RawData>;
+  url: string;
+  parser: (data: RawData) => Data;
+}): ApiResult<Data> => {
+  const [result, setResult] = useState<ApiResult<Data>>({ status: "loading" });
+
+  useEffect(() => {
+    fetch(url)
+      .then((response) => {
+        if (response.status === 200) {
+          return response.json();
+        } else if (response.status === 401) {
+          reload();
+
+          throw new Error("Unauthenticated");
+        } else {
+          // TODO: handle various errors
```

Comment:
In the meantime, could we throw an `Unhandled response code` error or something?

Reply:
Good point. I've made some improvements in 6f8d9b8. Note that throwing inside a promise just causes it to be rejected and the `throw` doesn't bubble up (which, no pun intended, threw me off when I was writing tests).